tests/unit/test_index.py

```diff
@@ -25,13 +26,99 @@
     group_locations,
 )
 from pip._internal.models.candidate import InstallationCandidate
+from pip._internal.models.index import PyPI
 from pip._internal.models.link import Link
 from pip._internal.models.search_scope import SearchScope
 from pip._internal.models.selection_prefs import SelectionPreferences
 from pip._internal.models.target_python import TargetPython
 from pip._internal.pep425tags import get_supported
 from pip._internal.utils.hashes import Hashes
-from tests.lib import CURRENT_PY_VERSION_INFO
+from pip._internal.utils.typing import MYPY_CHECK_RUNNING
+from tests.lib import CURRENT_PY_VERSION_INFO, make_test_search_scope
+
+if MYPY_CHECK_RUNNING:
+    from typing import List, Optional
+
+
+def make_fake_html_page(url):
+    html = dedent(u"""\
+    <html><head><meta name="api-version" value="2" /></head>
+    <body>
+    <a href="/abc-1.0.tar.gz#md5=000000000">abc-1.0.tar.gz</a>
+    </body></html>
+    """)
+    content = html.encode('utf-8')
+    headers = {}
+    return HTMLPage(content, url=url, headers=headers)
+
+
+def make_test_link_collector(
+    find_links=None,  # type: Optional[List[str]]
+):
+    # type: (...) -> LinkCollector
+    """
+    Create a LinkCollector object for testing purposes.
+    """
+    session = PipSession()
+    search_scope = make_test_search_scope(
+        find_links=find_links,
+        index_urls=[PyPI.simple_url],
+    )
+
+    return LinkCollector(
+        session=session,
+        search_scope=search_scope,
+    )
+
+
+def check_links_include(links, names):
+    """
+    Assert that the given list of Link objects includes, for each of the
+    given names, a link whose URL has a base name matching that name.
+    """
+    for name in names:
+        for link in links:
```

Comment:
I would do something like

```python
assert any(
    link.url.endswith(name) for link in links
), message_from_exception_below
```

Reply:
I wouldn’t use assert here, [this is not what it’s for](https://medium.com/@crystelpheonix/python-dont-use-assert-for-data-evaluation-721bfa93c571).

I’d prefer

```python
for name in names:
    if not any(link.url.endswith(name) for link in links):
        raise RuntimeError(...)
```

but Chris’s version is quite readable for me as well.

Reply:
I wouldn’t use assert here, Pytest can emit preplexing errors for asserts in functions. Personally I’d write a custom assertion, but Chris’s version would work well enough for me.

Maybe a middle of the road alternative would be satisfactory:

```python
for name in names:
    if not any(link.url.endswith(name) for link in links):
        raise RuntimeError(...)
```

Reply:
> Pytest can emit perplexing errors for asserts in functions

Do you have an example? I've used asserts in helper functions, classes, and fixtures and haven't noticed any issues so far.